src/lib/contractAddress.ts

```diff
@@ -0,0 +1,115 @@
+import { BigNumberish, ethers, keccak256 } from 'ethers';
+import { UnsafeApi, AccountId } from "polkadot-api"
+import { fromHex } from "polkadot-api/utils";
+
+/**
+ * Calculates the expected Ethereum-style contract address using CREATE or CREATE2 logic.
+ */
+export async function calculateContractAddress(
+    api: UnsafeApi<any>,
+    accountId: string,
+    code: string, 
+    data: Uint8Array,
+    salt: string,
+  ): Promise<string | undefined> {
+    let calculatedAddress: string | undefined = undefined;
+    const acIdEnc = AccountId()[0];
+    const mappedAccount = toEthAddress(acIdEnc(accountId));
+    if (salt) {
+        // Use CREATE2
+        calculatedAddress = create2(mappedAccount, fromHex(code), data, salt);
+    } else {
+        // @ts-ignore
+        const nonce = await api.apis.AccountNonceApi.account_nonce(accountId);
+         // Use CREATE1
+        if (nonce !== null) {
+            calculatedAddress = create1(mappedAccount, nonce);
+            console.log("calculatedAddress", calculatedAddress);
+        }
+    }
+    return calculatedAddress;
+}
+
+/**
```

Comment:
The code below is adapted from Contracts UI, with only modifications to use PAPI library: https://github.com/use-ink/contracts-ui/blob/426b6fddbd1fd2f409f2937f2ef8e3cab0e1c921/src/ui/hooks/useNewContract.ts#L160C6-L179C8